### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by c4pt0r**

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (ds *DataSource) Query(query string, args ...interface{}) (*sql.Rows, error) {
+	var err error
+	var ret *sql.Rows
+	if ds.db == nil {
+		cleanExit(errors.New("db is nil"))
```
Comment (ngaut) [inline db.go]: Return error.

File: db.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+			ds.db.Close()
+			if err := ds.Connect(); err != nil {
+				return nil, errors.Trace(err)
+			}
+			continue
+		} else {
+			return ret, nil
+		}
+	}
+	return nil, errors.Trace(err)
```
Comment (ngaut) [inline db.go]: return nil, nil

File: db.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		cleanExit(errors.New("db is nil"))
+	}
+
+	for i := 0; i < MaxRetryNum; i++ {
+		ret, err = ds.db.Query(query, args...)
+		if err != nil {
+			time.Sleep(500 * time.Millisecond)
+			ds.db.Close()
+			if err := ds.Connect(); err != nil {
+				return nil, errors.Trace(err)
```
Comment (ngaut) [inline db.go]: Functions are simple, could you remove all Trace call?

**Comment by c4pt0r** [comment]
@ngaut PTAL

File: main.go [hunk 4]
```diff
@@ -26,12 +27,26 @@ var (
 	count = flag.Int("n", 50, "Number of process to show")
 )
 
+func InitDB() error {
+	globalDS = newDataSource(*user, *pwd, *host, *port)
+	if err := globalDS.Connect(); err != nil {
+		return errors.Trace(err)
```
Comment (ngaut) [inline main.go]: No need to call Trace.

File: main.go [hunk 5]
```diff
@@ -26,12 +27,26 @@ var (
...
+	}
+	return nil
+}
+
 func main() {
 	flag.Parse()
+	if err := InitDB(); err != nil {
+		panic(err)
```
Comment (ngaut) [inline main.go]: Use CleanExit instead of panic.

File: main.go [hunk 6]
```diff
@@ -26,12 +27,26 @@ var (
...
+	}
 	if err := ui.Init(); err != nil {
 		panic(err)
```
Comment (ngaut) [inline main.go]: Ditto.

File: ui.go [hunk 7]
```diff
@@ -35,10 +35,56 @@ func (pg *ProcessListGrid) SetText(str string) {
...
+
+type HotSpotGrids struct {
+	grids []*ui.Grid
+}
+
+func newHotSpotGrids() *HotSpotGrids {
+	termWidth, _ := ui.TerminalDimensions()
+	ret := &HotSpotGrids{}
+	offset := 0
+	for i := 0; i < 5; i++ {
```
Comment (ngaut) [inline ui.go:54]: Could you explain the magic number 5?

Comment (c4pt0r) [reply ui.go:54]: I want to show top10 hot regions, which means we need 5 rows with 2 columns.

**Review by ngaut — APPROVED** [review]

**Pull request merged**